bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java

```diff
@@ -13,125 +13,179 @@
 package org.openhab.automation.jsscripting.internal.threading;
 
 import java.time.ZonedDateTime;
-import java.util.concurrent.TimeUnit;
+import java.util.HashMap;
 
-import org.eclipse.jdt.annotation.NonNullByDefault;
-import org.eclipse.jdt.annotation.Nullable;
 import org.openhab.core.model.script.ScriptServiceUtil;
-import org.openhab.core.model.script.actions.Timer;
 import org.openhab.core.scheduler.ScheduledCompletableFuture;
 import org.openhab.core.scheduler.Scheduler;
-import org.openhab.core.scheduler.SchedulerRunnable;
 
 /**
- * A replacement for the timer functionality of {@link org.openhab.core.model.script.actions.ScriptExecution
- * ScriptExecution} which controls multithreaded execution access to the single-threaded GraalJS contexts.
+ * A polyfill implementation of NodeJS timer functionality (<code>setTimeout()</code>, <code>setInterval()</code> and
+ * the cancel methods) which controls multithreaded execution access to the single-threaded GraalJS contexts.
  *
  * @author Florian Hotze - Initial contribution
+ * @author Florian Hotze - Reimplementation to conform standard JS setTimeout and setInterval
  */
 public class ThreadsafeTimers {
     private final Object lock;
+    private final Scheduler scheduler;
+    // Mapping of positive, non-zero integer values (used as timeoutID or intervalID) and the Scheduler
+    private final HashMap<Integer, ScheduledCompletableFuture<Object>> idSchedulerMapping = new HashMap<>();
+    private String identifier = "noIdentifier";
 
     public ThreadsafeTimers(Object lock) {
         this.lock = lock;
+        this.scheduler = ScriptServiceUtil.getScheduler();
     }
 
-    public Timer createTimer(ZonedDateTime instant, Runnable callable) {
-        return createTimer(null, instant, callable);
+    /**
+     * Set the identifier base string used for naming scheduled jobs.
+     *
+     * @param identifier identifier to use
+     */
+    public void setIdentifier(String identifier) {
+        this.identifier = identifier;
     }
 
-    public Timer createTimer(@Nullable String identifier, ZonedDateTime instant, Runnable callable) {
-        Scheduler scheduler = ScriptServiceUtil.getScheduler();
+    /**
+     * Creates a new and unused timerId.
+     *
+     * @return an unused timerId
+     */
+    private Integer createNewId() {
+        Integer id = idSchedulerMapping.size() + 1;
```

Comment:
Should we be reusing ids? What does JS do? I ask in case it always guarantees unused ids, in which case scripts could depend on this and reuse may break them.

Reply:
I just had a look at the mdn web docs, we should not:

> It is guaranteed that a timeoutID value will never be reused by a subsequent call to setTimeout() or setInterval() on the same object (a window or a worker). However, different objects use separate pools of IDs.